### PR TITLE
Configure TCP or UNIX socket

### LINKED ISSUE #5: Problem: The clamd socket type cannot be configured
The Clamav daemon can be configured to use a UNIX file or TCP socket but this role hasn't this option available when deploying.

### PR DESCRIPTION
**Pull request by mamedin**
A new variable `CLAMAV_SOCKET_TYPE` has been added. It can take two
values: "tcp" or "file".

On CentOS and Ubuntu 14 it is configured in the `clamd.conf` file.

On Ubuntu 16 this configuration is more complex because the
clamav-daemon package uses three systemd files to configure the socket
type (unix file or tcp socket). This PR uses debconf-set-selections to
configure the socket type, but the clamav-daemon package doesn't read
the debconf system when upgrading or reinstalling, it is used only when
installing from scratch, so a package purge is needed when changing the
socket type.

The SELinux policies `antivirus_can_scan_system` and `clamd_use_jit` will
be activated if SELinux is enabled on CentOS.

On Ubuntu task files, the "Enable services" task has been modified to
restart the clamav-daemon service instead of starting. It is needed to
apply the changes made by the role.

On Ubuntu the default socket type is UNIX file.
On Centos the default socket type is TCP.

Finally, 2 additional variables have been added to define the TCP port and IP address to listen on:

- CLAMAV_TCP_ADDR: Default 127.0.0.1
- CLAMAV_TCP_PORT: Default 3310

It Fixes #5

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -3,8 +3,15 @@ ansible-clamav
 
 Installs clamav (daemon and client). 
 
+Role Variables
+______________
+
+- `CLAMAV_SOCKET_TYPE`: configure clamad to use a tcp or file socket (default: "file")
+- `CLAMAV_MAX_FILE_SIZE`: (default: "2000M")
```
Comment (scollazo) [inline README.md]: Those fields should also have an explanation, not only `CLAMAV_SOCKET_TYPE`

File: README.md [hunk 2]
```diff
@@ -3,8 +3,15 @@ ansible-clamav
 
 Installs clamav (daemon and client). 
 
+Role Variables
+______________
+
+- `CLAMAV_SOCKET_TYPE`: configure clamad to use a tcp or file socket (default: "file")
```
Comment (scollazo) [inline README.md]: It reads `clamad`, typo?

File: defaults/main.yml [hunk 3]
```diff
@@ -5,3 +5,9 @@ CLAMAV_MAX_FILE_SIZE: "2000M"
 CLAMAV_MAX_SCAN_SIZE: "2000M"
 CLAMAV_MAX_STREAM_LENGTH: "2000M"
 
+# The clamvd can be configured to work with a file socket or a tcp socket.
+# Available Options for CLAMAV_SOCKET_TYPE are "tcp" and "file".
```
Comment (scollazo) [inline defaults/main.yml]: `clamvd`?

File: defaults/main.yml [hunk 4]
```diff
@@ -5,3 +5,9 @@ CLAMAV_MAX_FILE_SIZE: "2000M"
...
+# This env var has not a default value. 
+# On CentOS the role will use "tcp" if not defined
```
Comment (scollazo) [inline defaults/main.yml]: Itemize this?

```
#  - On CentOS...
#  - On Ubuntu...
```

File: tasks/main.yml [hunk 5]
```diff
@@ -5,7 +5,13 @@
   include: "ubuntu.yml"
   static: "no"
   when:
-    - ansible_distribution == "Ubuntu"
+    - ansible_distribution == "Ubuntu" and ansible_distribution_release != "trusty"
```
Comment (scollazo) [inline tasks/main.yml]: May be we should be explicit here, and use when `ansible_distribution_release= "xenial"`. Did you test this with newer Ubuntu versions (18.04 is almost out!)

File: tasks/redhat.yml [hunk 6]
```diff
@@ -54,9 +73,17 @@
...
-- name: "Enable and start services"
+- name: "Create tmpfiles custom socket directory"
+  shell: "echo 'd /var/run/clamav 0755 clamscan clamscan' > /etc/tmpfiles.d/clamd.scam.archivematica.conf"
+  when: CLAMAV_SOCKET_TYPE ==  'file'
+
+- name: "Run systemd-tmpfiles to create custom socket directory"
+  shell: "systemd-tmpfiles --create /etc/tmpfiles.d/clamd.scam.archivematica.conf"
+  when: CLAMAV_SOCKET_TYPE ==  'file'
+
+- name: "Enable and restart services"
```
Comment (scollazo) [inline tasks/redhat.yml:84]: This should be triggered only when a configuration change is made

File: tasks/ubuntu.yml [hunk 7]
```diff
@@ -7,6 +7,66 @@
...
+  when:  CLAMAV_SOCKET_TYPE is not defined
+
+- set_fact:
+    clamav_daemon_already_installed: true
+  when:  dpkg_check_clamav_daemon.rc == 0
+
+# clamav-daemon already installed block
+- block:
+
+   # check if clamav is running with file or tcp socket
```
Comment (scollazo) [inline tasks/ubuntu.yml]: First letter should be capital

File: tasks/ubuntu.yml [hunk 8]
```diff
@@ -7,6 +7,66 @@
...
+     register: clamavunixsocket
+     check_mode: no
+     ignore_errors: yes
+     changed_when: no
+
+   - set_fact:
+       clamav_uses_file_socket: true
+     when: clamavunixsocket is defined and ( clamavunixsocket.rc == 0 )
+
+   # purge clamav-daemon if was installed and socket type will change
```
Comment (scollazo) [inline tasks/ubuntu.yml]: First letter should be capital

File: tasks/ubuntu.yml [hunk 9]
```diff
@@ -7,6 +7,66 @@
...
+       name: clamav-daemon
+       purge: yes
+       state: absent
+       force: yes
+     when: 
+       - clamav_daemon_already_installed is defined
+       - CLAMAV_SOCKET_TYPE !=  'tcp' and clamav_uses_file_socket is not defined or clamav_daemon_already_installed is defined and CLAMAV_SOCKET_TYPE == 'tcp' and clamav_uses_file_socket is defined
+  when: clamav_daemon_already_installed is defined
+
+# set debconf when installing clamav-daemon
```
Comment (scollazo) [inline tasks/ubuntu.yml]: ...you know what I'm about to say here...

File: tasks/ubuntu.yml [hunk 10]
```diff
@@ -39,7 +99,7 @@
 - name: "Enable services"
```
Comment (scollazo) [inline tasks/ubuntu.yml:101]: As with Centos, this should be triggered only on configuration change

File: templates/etc/clamd.d/scan.conf.j2 [hunk 11]
```diff
@@ -99,15 +104,23 @@ LogSyslog yes
...
+{% endif %}
 
 # TCP address.
 # By default we bind to INADDR_ANY, probably not wise.
 # Enable the following to provide some degree of protection
 # from the outside world. This option can be specified multiple
 # times if you want to listen on multiple IPs. IPv6 is now supported.
 # Default: no
+{% if CLAMAV_SOCKET_TYPE == 'file' %}
+#TCPAddr 127.0.0.1
```
Comment (scollazo) [inline templates/etc/clamd.d/scan.conf.j2]: Can the tcp bind address and port be converted to vars? There are environments where we need to have clamav listening on a different port, due to 3310 being used by other service.

**Review by scollazo — CHANGES_REQUESTED** [review]

**Comment by mamedin** [comment]
@scollazo  thanks for your suggestions. I have changed all that you said except triggering the clamav-daemon restart only on configuration changes. I think I have added a lot of complexity to this role, especially in ubuntu xenial and the clamd is a very light demon that can be restarted without any problem.

**Review by sevein — APPROVED** [review]

**Review by hakamine — APPROVED** [review]
:+1: I added a commit with travis config to run a syntax check on the role, and do a test run with the default configuration, on the currently 3 supported OSes, tests are passing.

**Comment by sevein** [comment]
> 👍 I added a commit with travis config to run a syntax check on the role, and do a test run with the default configuration, on the currently 3 supported OSes, tests are passing.

🎉 this is interesting, @hakamine. So there's not test assertions other than expecting that the playbook completes successfully, right? I guess that's a good start.

**Pull request merged**